core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -265,17 +277,32 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
             return allocationService.applyFailedShards(currentState, failedShards);
         }
 
-        private boolean shardExists(ClusterState currentState, ShardRoutingEntry task) {
+        private enum TaskClassification {
+            EXISTS_LEGAL,
+            EXISTS_ILLEGAL,
+            DOES_NOT_EXIST
+        }
+
+        private TaskClassification classifyTask(ClusterState currentState, ShardRoutingEntry task) {
             RoutingNodes.RoutingNodeIterator routingNodeIterator =
                 currentState.getRoutingNodes().routingNodeIter(task.getShardRouting().currentNodeId());
             if (routingNodeIterator != null) {
                 for (ShardRouting maybe : routingNodeIterator) {
                     if (task.getShardRouting().isSameAllocation(maybe)) {
-                        return true;
+                        IndexShardRoutingTable indexShard =
+                            currentState.getRoutingTable().shardRoutingTable(task.getShardRouting().index().getName(), task.getShardRouting().getId());
+                        ShardRouting primaryShard = indexShard.primaryShard();
+                        if (task.identity.allocationId().equals(maybe.allocationId())) {
```

Comment:
can we give precedence to the source shard check? if a stale primary wants to fail a shard that's not there, we should still fail the operation so that the stale primary will fail itself.